server/pseudonymsys.go

```diff
@@ -39,41 +39,52 @@ func (s *Server) PseudonymsysGenerateNym(req *pb.Message, stream pb.Protocol_Run
 	blindedB := new(big.Int).SetBytes(proofRandData.B2)
 	signatureR := new(big.Int).SetBytes(proofRandData.R)
 	signatureS := new(big.Int).SetBytes(proofRandData.S)
+	regKey := proofRandData.RegKey
+
+	regKeyOk, err := s.registrationManager.CheckRegistrationKey(regKey)
 
-	challenge, err := org.GetChallenge(nymA, blindedA, nymB, blindedB, x1, x2, signatureR, signatureS)
 	var resp *pb.Message
-	if err != nil {
+
+	if !regKeyOk || err != nil {
 		resp = &pb.Message{
-			Content: &pb.Message_PedersenDecommitment{
-				&pb.PedersenDecommitment{},
-			},
-			ProtocolError: err.Error(),
+			Content: nil,
+			ProtocolError: "Registration key verification failed",
 		}
 	} else {
-		resp = &pb.Message{
-			Content: &pb.Message_PedersenDecommitment{
-				&pb.PedersenDecommitment{
-					X: challenge.Bytes(),
+		challenge, err := org.GetChallenge(nymA, blindedA, nymB, blindedB, x1, x2, signatureR, signatureS)
+		if err != nil {
+			resp = &pb.Message{
+				Content: &pb.Message_PedersenDecommitment{
+					&pb.PedersenDecommitment{},
 				},
-			},
+				ProtocolError: err.Error(),
+			}
+		} else {
+			resp = &pb.Message{
+				Content: &pb.Message_PedersenDecommitment{
+					&pb.PedersenDecommitment{
+						X: challenge.Bytes(),
+					},
+				},
+			}
 		}
-	}
 
-	if err := s.send(resp, stream); err != nil {
-		return err
-	}
+		if err := s.send(resp, stream); err != nil {
+			return err
+		}
 
-	req, err = s.receive(stream)
-	if err != nil {
-		return err
-	}
+		req, err = s.receive(stream)
+		if err != nil {
+			return err
+		}
 
-	proofData := req.GetSchnorrProofData() // SchnorrProofData is used in DLog equality proof as well
-	z := new(big.Int).SetBytes(proofData.Z)
-	valid := org.Verify(z)
+		proofData := req.GetSchnorrProofData() // SchnorrProofData is used in DLog equality proof as well
+		z := new(big.Int).SetBytes(proofData.Z)
+		valid := org.Verify(z)
 
-	resp = &pb.Message{
-		Content: &pb.Message_Status{&pb.Status{Success: valid}},
+		resp = &pb.Message{
+			Content: &pb.Message_Status{&pb.Status{Success: valid}},
+		}
 	}
 
 	if err = s.send(resp, stream); err != nil {
```

Comment:
I don't have strong preference here, but it might be more readable to call send in both if !regKeyOk block and else block. Otherwise the same send is used for the first (if !regKeyOk) and last (else) reply. But yes, this would make the code longer.

Reply:
No strong preference either, so I will agree with you here.

TBH, I asked @mancabizjak and she is leaning towards your option :)